core/src/sumeragi/mod.rs

```diff
@@ -105,30 +76,43 @@ impl Sumeragi {
     ///
     /// # Panics
     /// - If either mutex is poisoned
-    #[allow(
-        clippy::expect_used,
-        clippy::unwrap_in_result,
-        clippy::cast_precision_loss,
-        clippy::float_arithmetic,
-        clippy::mutex_integer
-    )]
+    #[allow(clippy::cast_precision_loss)]
     pub fn update_metrics(&self) -> Result<()> {
         let online_peers_count: u64 = self
-            .internal
             .network
             .online_peers(std::collections::HashSet::len)
             .try_into()
             .expect("casting usize to u64");
 
-        let wsv_guard = self.internal.wsv.lock();
+        let (
+            height,
+            domains,
+            genesis_timestamp,
+            metric_tx_amounts_counter,
+            metric_tx_amounts,
+            latest_block_view_change_index,
+        ) = self.wsv(|wsv| {
+            (
+                wsv.height(),
+                // Not very nice to clone, but this way we don't hold lock
+                wsv.domains()
+                    .iter()
+                    .map(|domain_ref| (domain_ref.key().clone(), domain_ref.value().accounts.len()))
+                    .collect::<Vec<_>>(),
+                wsv.genesis_timestamp(),
+                wsv.metric_tx_amounts_counter.get(),
+                wsv.metric_tx_amounts.get(),
+                wsv.latest_block_view_change_index(),
+            )
+        });
```

Comment:
Probably deserves being its own function. 

Also, while I understand why you're doing this micro-optimisation, unless you know that the locks are contended, the allocator pressure is more important. Typically if you clone a large structure, the compiler can hint to the allocator that if it, it should look for a similarly sized chunk. If you do the cloning manually, the compiler might not be smart enough to figure it out on its own, and create highly fragmented memory. 

What's more, the real solution to the issue probably has to do more with fixing the amount of data and locality of said data in the WSV. If it is a differential structure, the mutex is no longer necessary and can be replaced with cheap clones.

Reply:
> Imo we shouldn't extract this into separate function since it's used only once and it's just bunch of wsv data access operations under one lock.

The function where this belongs is already quite hard to read. If this were a `fn` item returning a struct, this entire block could be relocated elsewhere out of sight. Not exactly necessary, but you yourself remarked that this block was ugly. This is one way to improve its aesthetics.